File: backend/app/schemas/public_voice_calls.py
from __future__ import annotations

from typing import Literal

from pydantic import BaseModel, ConfigDict, Field


class PublicVoiceCallRequest(BaseModel):
    model_config = ConfigDict(extra="forbid", strict=True)
    context_token: str = Field(min_length=32, max_length=512)


class PublicVoiceCallCapabilities(BaseModel):
    submissions: bool = True
    calls: bool = True


class PublicVoiceCallResponse(BaseModel):
    status: Literal["ready"]
    join_url: str
    capabilities: PublicVoiceCallCapabilities = PublicVoiceCallCapabilities()


class PublicVoiceCallbackResponse(BaseModel):
    status: Literal["accepted"]


PublicCallErrorCode = Literal[
    "experience_unavailable",
    "experience_version_changed",
    "call_already_started",
    "call_state_conflict",
    "context_session_unavailable",
    "context_session_expired",
    "validation_error",
    "rate_limited",
    "call_unavailable",
    "call_provider_unavailable",
    "phone_unavailable",
    "destination_not_allowed",
    "call_capacity_reached",
    "internal_error",
]


class PublicVoiceCallError(BaseModel):
    code: PublicCallErrorCode

File: frontend/lib/api/public-voice-calls.ts
export type PublicCallErrorCode =
  | 'experience_unavailable'
  | 'experience_version_changed'
  | 'call_already_started'
  | 'call_state_conflict'
  | 'context_session_unavailable'
  | 'context_session_expired'
  | 'validation_error'
  | 'rate_limited'
  | 'call_unavailable'
  | 'call_provider_unavailable'
  | 'phone_unavailable'
  | 'destination_not_allowed'
  | 'call_capacity_reached'
  | 'internal_error';

export type PublicCallResult =
  | { ok: true; data: { status: 'ready'; join_url: string } }
  | { ok: false; error: PublicCallErrorCode };

export async function launchPublicVoiceCall(slug: string, contextToken: string): Promise<PublicCallResult> {
  const baseUrl = process.env.NEXT_PUBLIC_API_URL || 'http://127.0.0.1:8000';
  try {
    const response = await fetch(`${baseUrl}/api/v1/public/voice-experiences/${encodeURIComponent(slug)}/calls`, {
      method: 'POST',
      cache: 'no-store',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify({ context_token: contextToken }),
    });
    const body = await response.json().catch(() => ({}));
    if (!response.ok) return { ok: false, error: body?.detail?.code || 'internal_error' };
    return { ok: true, data: body };
  } catch {
    return { ok: false, error: 'internal_error' };
  }
}

export async function requestPublicVoiceCallback(
  slug: string,
  contextToken: string,
): Promise<{ ok: true } | { ok: false; error: PublicCallErrorCode }> {
  const baseUrl = process.env.NEXT_PUBLIC_API_URL || 'http://127.0.0.1:8000';
  try {
    const response = await fetch(
      `${baseUrl}/api/v1/public/voice-experiences/${encodeURIComponent(slug)}/callback-requests`,
      {
        method: 'POST',
        cache: 'no-store',
        headers: { 'Content-Type': 'application/json' },
        body: JSON.stringify({ context_token: contextToken }),
      },
    );
    const body = await response.json().catch(() => ({}));
    if (!response.ok) return { ok: false, error: body?.detail?.code || 'internal_error' };
    return { ok: true };
  } catch {
    return { ok: false, error: 'internal_error' };
  }
}
